subprojects/xcfa/xcfa-analysis/src/main/java/hu/bme/mit/theta/xcfa/analysis/XcfaPrecRefiner.kt

```diff
@@ -41,18 +41,9 @@ class XcfaPrecRefiner<S : ExprState, P : Prec, R : Refutation>(refToPrec: Refuta
         Preconditions.checkNotNull<R>(refutation)
         var runningPrec: P = prec.p
         for (i in trace.states.indices) {
-            val reverseLookup = trace.states[i].processes.values.map {
-                it.varLookup.map {
-                    it.map {
-                        Pair(it.value, it.key)
-                    }
-                }.flatten()
-            }.flatten().toMap()
-            val additionalLookup = if (i > 0) getTempLookup(
-                trace.actions[i - 1].edge.label).entries.associateBy(
-                { it.value }) { it.key } else emptyMap()
-            val precFromRef = refToPrec.toPrec(refutation, i)
-                .changeVars(reverseLookup + additionalLookup)
+            val tempLookup = if (i > 0) getTempLookup(trace.actions[i - 1].edge.label).entries
+                .associateBy({ it.value }) { it.key } else emptyMap()
+            val precFromRef = refToPrec.toPrec(refutation, i).changeVars(tempLookup)
```

Comment:
The problem with the non-terminating multithreaded test is caused by removal of general parameters from the precision here (`reverseLookup` was deleted). This was necessary for recursive tasks, where the concrete variable _instances of different procedure calls_ are needed in the precision in order to successfully verify them. This seems to cause problem for multithreaded tasks though, where we have _instances of different threads_. As a quick fix, we could omit reverseLookup when verifyingrecursive tasks and use it otherwise. We should discuss how to proceed on the long-term.

Reply:
Thanks for looking into it! I think we should discuss the next steps later, in person.